File: rd-cpp/src/rd_core_cpp/src/test/cases/AdviseVsViewTest.cpp
//
// Created by jetbrains on 13.07.2018.
//


#include "gtest/gtest.h"

#include "LifetimeDefinition.h"
#include "Property.h"

using namespace rd;

TEST(advise_vs_view, advise_behaviour1) {
    //int* p = new int(1);
    LifetimeDefinition lifetimeDef(Lifetime::Eternal());
    Property<bool> property(false);
    Lifetime lifetime = lifetimeDef.lifetime;

    std::vector<bool> log;

    lifetime->add_action([&property]() { property.set(true); });
    property.advise(lifetime, [&log](bool const &value) {
        log.push_back(value);
    });
    lifetimeDef.terminate();

    std::vector<bool> expected{false};
    EXPECT_EQ(expected, log);
}

TEST(advise_vs_view, view_behaviour1) {
    LifetimeDefinition lifetimeDef(Lifetime::Eternal());
    Property<bool> property(false);
    Lifetime lifetime = lifetimeDef.lifetime;

    std::vector<bool> log;

    lifetime->add_action([&property]() { property.set(true); });
    property.view(lifetime, [&log](Lifetime _, bool const &value) {
        log.push_back(value);
    });
//	int c = lifetime.use_count();
    lifetimeDef.terminate();

    std::vector<bool> expected{false};
    EXPECT_EQ(expected, log);
}

TEST(advise_vs_view, advise_behaviour2) {
    LifetimeDefinition lifetimeDef(Lifetime::Eternal());
    Property<bool> property(false);
    Lifetime lifetime = lifetimeDef.lifetime;

    std::vector<bool> log;


    property.advise(lifetime, [&log](bool const &value) {
        log.push_back(value);
    });
    lifetime->add_action([&property]() { property.set(true); });
    lifetimeDef.terminate();

    std::vector<bool> expected{false};
    EXPECT_EQ(expected, log);
}

TEST(advise_vs_view, view_behaviour2) {
    LifetimeDefinition lifetimeDef(Lifetime::Eternal());
    Property<bool> property(false);
    Lifetime lifetime = lifetimeDef.lifetime;

    std::vector<bool> log;

    property.view(lifetime, [&log](Lifetime _, bool const &value) {
        log.push_back(value);
    });
    lifetime->add_action([&property]() { property.set(true); });
    lifetimeDef.terminate();

    std::vector<bool> expected{false, true};
    EXPECT_EQ(expected, log);
}

TEST(advise_vs_view, advise_behaviour3) {
    LifetimeDefinition lifetimeDef(Lifetime::Eternal());
    Property<int32_t> property_a(0);
    Property<int32_t> property_b(0);

    Lifetime lifetime = lifetimeDef.lifetime;

    std::vector<int> log_a;
    std::vector<int> log_b;


    property_a.advise(lifetime, [&log_a](int const &value) {
        log_a.push_back(value);
    });

    property_b.advise(lifetime, [&log_b](int const &value) {
        log_b.push_back(value);
    });

    property_a.set(1);
    property_b.set(2);

    lifetimeDef.terminate();

    property_a.set(3);
    property_b.set(4);

    std::vector<int> expected_a{0, 1};
    std::vector<int> expected_b{0, 2};
    EXPECT_EQ(expected_a, log_a);
    EXPECT_EQ(expected_b, log_b);
}

TEST(advise_vs_view, view_behaviour3) {
    LifetimeDefinition lifetimeDef(Lifetime::Eternal());
    Property<int32_t> property_a(0);
    Property<int32_t> property_b(0);

    Lifetime lifetime = lifetimeDef.lifetime;

    std::vector<int> log_a;
    std::vector<int> log_b;


    property_a.view(lifetime, [&log_a](Lifetime _, int const &value) {
        log_a.push_back(value);
    });

    property_b.view(lifetime, [&log_b](Lifetime _, int const &value) {
        log_b.push_back(value);
    });

    property_a.set(1);
    property_b.set(2);

    lifetimeDef.terminate();

    property_a.set(3);
    property_b.set(4);

    std::vector<int> expected_a{0, 1};
    std::vector<int> expected_b{0, 2};
    EXPECT_EQ(expected_a, log_a);
    EXPECT_EQ(expected_b, log_b);
}

File: rd-core/src/commonTest/kotlin/com/jetbrains/rd/util/test/cases/AdviseVsViewTest.kt
package com.jetbrains.rd.util.test.cases

import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import com.jetbrains.rd.util.lifetime.onTermination
import com.jetbrains.rd.util.reactive.Property
import com.jetbrains.rd.util.test.framework.RdTestBase
import kotlin.test.Test
import kotlin.test.assertEquals

class AdviseVsViewTest : RdTestBase() {
    @Test
    fun adviseBehavior1() {
        val lifetimeDef = Lifetime.Eternal.createNested()
        val property = Property(false)
        val lifetime = lifetimeDef.lifetime

        val log = arrayListOf<Boolean>()

        lifetime.onTermination { property.set(true) }
        property.advise(lifetime) { /*lt, */value ->
            log.add(value)
        }
        lifetimeDef.terminate()

        assertEquals(listOf(false), log)
    }

    @Test
    fun viewBehavior1() {
        val lifetimeDef = Lifetime.Eternal.createNested()
        val property = Property(false)
        val lifetime = lifetimeDef.lifetime

        val log = arrayListOf<Boolean>()

        lifetime.onTermination { property.set(true) }
        property.view(lifetime) { _, value ->
            log.add(value)
        }
        lifetimeDef.terminate()

        assertEquals(listOf(false), log)
    }

    @Test
    fun adviseBehavior2() {
        val lifetimeDef = Lifetime.Eternal.createNested()
        val property = Property(false)
        val lifetime = lifetimeDef.lifetime

        val log = arrayListOf<Boolean>()

        property.advise(lifetime) { /*lt, */value ->
            log.add(value)
        }
        lifetime.onTermination { property.set(true) }
        lifetimeDef.terminate()

        assertEquals(listOf(false, true), log)
    }

    @Test
    fun viewBehavior2() {
        val lf = LifetimeDefinition()
        val property = Property(false)

        val log = arrayListOf<Boolean>()

        property.view(lf) { _, value ->
            // previously would throw an exception on viewing changes from { property.set(true) }
            log.add(value)
        }
        lf.onTermination { property.set(true) }
        lf.terminate()

        assertEquals(listOf(false), log)
    }

    @Test
    fun adviseBehavior3() {
        val lifetimeDef = Lifetime.Eternal.createNested()
        val propertyA = Property(0)
        val propertyB = Property(0)
        val lifetime = lifetimeDef.lifetime

        val logA = arrayListOf<Int>()
        val logB = arrayListOf<Int>()

        propertyA.advise(lifetime) { value -> logA.add(value) }
        propertyB.advise(lifetime) { value -> logB.add(value) }

        propertyA.set(1)
        propertyB.set(2)

        lifetimeDef.terminate()

        propertyA.set(3)
        propertyB.set(4)

        assertEquals(listOf(0, 1), logA)
        assertEquals(listOf(0, 2), logB)
    }

    @Test
    fun viewBehavior3() {
        val lifetimeDef = Lifetime.Eternal.createNested()
        val propertyA = Property(0)
        val propertyB = Property(0)
        val lifetime = lifetimeDef.lifetime

        val logA = arrayListOf<Int>()
        val logB = arrayListOf<Int>()

        propertyA.view(lifetime) { _, value -> logA.add(value) }
        propertyB.view(lifetime) { _, value -> logB.add(value) }

        propertyA.set(1)
        propertyB.set(2)

        lifetimeDef.terminate()

        propertyA.set(3)
        propertyB.set(4)

        assertEquals(listOf(0, 1), logA)
        assertEquals(listOf(0, 2), logB)
    }
}